src/H5Iint.c

```diff
@@ -9478,6 +9162,19 @@ H5I__new_mt_id_info(hid_t id, unsigned count, unsigned app_count, const void * o
                     assert(atomic_load(&(id_info_ptr->on_fl)));
                     atomic_store(&(id_info_ptr->on_fl), FALSE);
 
+                    assert( 0 < atomic_load(&(id_info_ptr->serial_num ) ) );
+
+                    /* the above assert only exists in production builds.  If id_info_ptr->serial_num
+                     * is zero and we get this far, increment H5I_mt_g.num_id_info_fl_head_sn_is_zero.
+                     */
+                    if ( 0 == atomic_load(&(id_info_ptr->serial_num)) ) {
```

Comment:
Again, I think that finding that the head of the free list is flagged as not being in the free list probably warrants an assertion failure right after `fl_shead` is loaded.